src/util/expr.h

```diff
@@ -161,6 +162,126 @@ class exprt:public irept
 
   void visit(class expr_visitort &visitor);
   void visit(class const_expr_visitort &visitor) const;
+
+  // Depth first search iterator - iterates over supplied expression
+  // and all its operands recursively.
+  template<typename expr_t>
+  class depth_iteratort final:
+  public std::iterator<std::forward_iterator_tag, expr_t, void>
+  {
+  public:
+    // Create end iterator
+    depth_iteratort() { }
+    // Create begin iterator, starting at the supplied node
+    explicit depth_iteratort(expr_t &root)
+    {
+      static_assert(std::is_same<exprt,
+      typename std::remove_const<expr_t>::type>::value,
+      "Only exprt and const exprt are allowed");
+      this->push_expr(root);
+    }
+    bool operator==(const depth_iteratort &other) const
+    {
+      return m_stack.size()==other.m_stack.size()
+            && std::equal(m_stack.begin(),
+                          m_stack.end(),
+                          other.m_stack.begin());
+    }
+    bool operator!=(const depth_iteratort &other) const
+    {
+      return !(*this==other);
+    }
+    depth_iteratort &operator++()
+    {
+      for(;;)
+      {
+        if(m_stack.back().it==m_stack.back().end)
+        {
+          m_stack.pop_back();
+          if(m_stack.empty())
+          {
+            return *this;
+          }
+          m_stack.back().it++;
+        }
+        else
+        {
+          push_expr(*m_stack.back().it);
+          return *this;
+        }
+      }
+    }
+    expr_t &operator*()
+    {
+      return m_stack.back().expr.get();
+    }
+    const expr_t &operator*() const
+    {
+      return m_stack.back().expr.get();
+    }
+    expr_t *operator->()
+    {
+      return &**this;
+    }
+    const expr_t *operator->() const
+    {
+      return &**this;
+    }
+    private:
+    typedef typename std::conditional<
+      std::is_const<expr_t>::value,
+      std::vector<exprt>::const_iterator,
```

Comment:
std::vector<const exprt> does not exist - T has to be assignable. Const types are not. As well as references, which means I std::reference_wrapper is used here instead of a reference.